### PR TITLE
New Rule: requireArrowFunction

### LINKED ISSUE #1383: requireArrowFunctions (ES6)
Could probably make a better/ more descriptive name.

[airbnb/javascript#8.1](https://github.com/airbnb/javascript#8.1)

> When you must use function expressions (as when passing an anonymous function), use arrow function notation.

```js
 // bad
  [1, 2, 3].map(function (x) {
    return x * x;
  });

  // good
  [1, 2, 3].map((x) => {
    return x * x;
  });
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/15297901-requirearrowfunctions-es6?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by hzoo**
Fixes #1383

### CODE REVIEW

**Comment by mrjoelkemp** [comment]
Can we add a test for the multi-line form of an arrow function?

To clarify, this form: 

```js
foo => {
  console.log('hi');
  console.log('there);
}
```

File: test/specs/rules/require-arrow-functions.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+describe('rules/require-arrow-functions', function() {
+    var checker;
+
+    beforeEach(function() {
+        checker = new Checker();
+        checker.registerDefaultRules();
+        checker.configure({ requireArrowFunctions: true, esnext: true });
+    });
+
+    it('should report use of function expression in declaration', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/require-arrow-functions.js]: For sanity, can we add a test for function expressions assignments as object properties?

i.e., 

```js
var foo = {};
foo.bar = function() {};
```

and

```js
var foo = {
  bar: function() {}
};
```

Named function expressions

File: test/specs/rules/require-arrow-functions.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    it('should report use of function expression as callback', function() {
+        assert(checker.checkString([
+            'a.map(function(n) {',
+                'return n + 1;',
+            '});'
+        ].join('\n')).getErrorCount() === 1);
+    });
+
+    it('should not report a function declaration', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/require-arrow-functions.js]: Can we also add a test for named function expressions?

e.g., 

should still report on 

```js
var foo = function mySweetFunc() {};
```

**Comment by hzoo** [comment]
Ok nice @mrjoelkemp, based on your comments I added tests for getter/setters, shorthand object methods which failed so will have to add some logic for those.

**Comment by mrjoelkemp** [comment]
Nice work! LGTM

**Comment by markelog** [comment]
Hooray for the first ES6-only rule!

**Pull request closed without merging**